client/packages.go

```diff
@@ -31,41 +32,156 @@ import (
+	Private          bool
+	DevMode          bool
+	JailMode         bool
+	TryMode          bool
+	Apps             []AppInfo
+	Broken           string
+	Contact          string
+	License          string
+
+	Prices      map[string]float64
+	Screenshots []Screenshot
 
 	// The flattended channel map with $track/$risk
-	Channels map[string]*snap.ChannelSnapInfo `json:"channels,omitempty"`
+	Channels map[string]*snap.ChannelSnapInfo
 
 	// The ordered list of tracks that contains channels
-	Tracks []string                `json:"tracks,omitempty"`
+	Tracks []string
+}
+
+type marshalledSnap struct {
+	ID               string                           `json:"id"`
+	Title            string                           `json:"title,omitempty"`
+	Summary          string                           `json:"summary"`
+	Description      string                           `json:"description"`
+	DownloadSize     int64                            `json:"download-size,omitempty"`
+	Icon             string                           `json:"icon,omitempty"`
+	InstalledSize    int64                            `json:"installed-size,omitempty"`
+	InstallDate      *time.Time                       `json:"install-date,omitempty"`
+	Name             string                           `json:"name"`
+	Developer        string                           `json:"developer"`
+	Status           string                           `json:"status"`
+	Type             string                           `json:"type"`
+	Version          string                           `json:"version"`
+	Channel          string                           `json:"channel"`
+	TrackingChannel  string                           `json:"tracking-channel,omitempty"`
+	IgnoreValidation bool                             `json:"ignore-validation,omitempty"`
+	Revision         snap.Revision                    `json:"revision"`
+	Confinement      string                           `json:"confinement"`
+	Private          bool                             `json:"private,omitempty"`
+	DevMode          bool                             `json:"devmode,omitempty"`
+	JailMode         bool                             `json:"jailmode,omitempty"`
+	TryMode          bool                             `json:"trymode,omitempty"`
+	Apps             []AppInfo                        `json:"apps,omitempty"`
+	Broken           string                           `json:"broken,omitempty"`
+	Contact          string                           `json:"contact"`
+	License          string                           `json:"license,omitempty"`
+	Prices           map[string]float64               `json:"prices,omitempty"`
+	Screenshots      []Screenshot                     `json:"screenshots,omitempty"`
+	Channels         map[string]*snap.ChannelSnapInfo `json:"channels,omitempty"`
+	Tracks           []string                         `json:"tracks,omitempty"`
+}
+
+// MarshalJSON makes Snap a json.Marshaller
+func (s *Snap) MarshalJSON() ([]byte, error) {
```

Comment:
The usual way to handle such case this in Go is to do the following:
```
func (s *Snap) MarshalJSON() ([]byte, error) {
	type auxSnap Snap  // use auxiliary type so that Go does not call Snap.MarshalJSON()
	// separate type just for marshalling
	m := struct {
		auxSnap
		InstallDate *time.Time `json:"install-date,omitempty"`
	}{
		auxSnap: auxSnap(*s),
	}
	if !s.InstallDate.IsZero() {
		m.InstallDate = &s.InstallDate
	}
	return json.Marshal(&m)
}

// and you don't need to add unmarshaller anymore
```
See for an example: https://play.golang.org/p/nHK9bjfd_c

Reply:
Oh, that's much nicer. I copied the existing cases in snapd - they should be update too then.

Reply:
Thanks for updating this :)